server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1369,36 +1426,42 @@ public Metadata build() {
                 }
             }
 
-            aliasAndIndexLookup.values().stream()
-                .filter(aliasOrIndex -> aliasOrIndex.getType() == IndexAbstraction.Type.ALIAS)
+            indexAbstractionLookup.values().stream()
+                .filter(indexAbstraction -> indexAbstraction.getType() == IndexAbstraction.Type.ALIAS)
                 .forEach(alias -> ((IndexAbstraction.Alias) alias).computeAndValidateAliasProperties());
-            return aliasAndIndexLookup;
+            return indexAbstractionLookup;
         }
 
-        private void validateDataStreams(SortedMap<String, IndexAbstraction> indicesLookup) {
+        private void validateDataStreams(SortedSet<String> allIndices, SortedSet<String> allAliases) {
             DataStreamMetadata dsMetadata = (DataStreamMetadata) customs.get(DataStreamMetadata.TYPE);
             if (dsMetadata != null) {
                 for (DataStream ds : dsMetadata.dataStreams().values()) {
-                    IndexAbstraction existing = indicesLookup.get(ds.getName());
-                    if (existing != null && existing.getType() != IndexAbstraction.Type.DATA_STREAM) {
+                    if (allIndices.contains(ds.getName()) || allAliases.contains(ds.getName())) {
```

Comment:
Perhaps open a separate pr for the change to this method? This adds additional validation outside concrete index to data stream lookup.